drivers/frequency/adf4382/adf4382.c

```diff
@@ -0,0 +1,1199 @@
+ * 		     or negative error code otherwise.
+ */
+int adf4382_spi_test_bits(struct adf4382_dev *dev, uint16_t reg_addr,
+			  uint8_t bits)
+{
+	uint8_t val, ret;
+
+	ret = adf4382_spi_read(dev, reg_addr, &val);
+	if (ret)
+		return ret;
+
+	return (val & bits) == bits;
+}
+
+/**
+ * @brief Will output on the terminal the values of all the ADF4382 registers.
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
+	uint16_t i;
+
+	for(i = 0; i < 0x68; i++) {
+		ret = adf4382_spi_read(dev, i, &val);
+		if (ret)
+			return ret;
+
+		pr_info("0x%X    0x%X\n", i, val);
+	}
+
+	for(i = 0x100; i < 0x112; i++) {
+		ret = adf4382_spi_read(dev, i, &val);
+		if (ret)
+			return ret;
+
+		pr_info("0x%X    0x%X\n", i, val);
+	}
+
+	for(i = 0x200; i < 0x274; i++) {
+		ret = adf4382_spi_read(dev, i, &val);
+		if (ret)
+			return ret;
+
+		pr_info("0x%X    0x%X\n", i, val);
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Set the desired reference frequency and reset everything over to maximum
+ * supported value of 5GHz to the max. value and everything under the minimum
+ * supported value of 10MHz to the min. value.
+ * @param dev 		- The device structure.
+ * @param val		- The desired reference frequency in Hz.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_ref_clk(struct adf4382_dev *dev, uint64_t val)
+{
+	dev->ref_freq_hz = (uint64_t)val;
+
+	if(val > ADF4382_REF_CLK_MAX)
+		dev->ref_freq_hz = ADF4382_REF_CLK_MAX;
+
+	if(val < ADF4382_REF_CLK_MIN)
+		dev->ref_freq_hz = ADF4382_REF_CLK_MIN;
+
+	return adf4382_set_freq(dev);
+}
+
+/**
+ * @brief Gets the user proposed reference frequency
+ * @param dev 		- The device structure.
+ * @param val		- The set value of the reference frequency in Hz.
+ * @return    		- Reference frequency in KHz.
+ */
+void adf4382_get_ref_clk(struct adf4382_dev *dev, uint64_t *val)
+{
+	*val = dev->ref_freq_hz;
+}
+
+/**
+ * @brief Set the reference doubler to enable or disable based on the passed
+ * parameter. If the parameter is different then 0 it will set the doubler to
+ * enable.
+ * @param dev 		- The device structure.
+ * @param en	 	- The enable or disable value of the reference doubler.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_en_ref_doubler(struct adf4382_dev *dev, bool en)
+{
+	dev->ref_doubler_en = en;
+
+	return adf4382_set_freq(dev);
+}
+
+/**
+ * @brief Gets the value the doubler if it is enabled or disable and stores it
+ * it the dev structure.
+ * @param dev 		- The device structure.
+ * @param en		- The read value of the reference doubler.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_get_en_ref_doubler(struct adf4382_dev *dev, bool *en)
+{
+	int tmp;
+
+	tmp = adf4382_spi_test_bits(dev, 0x20, ADF4382_EN_RDBLR_MSK);
```

Comment:
is the "test bit" function required here. a getter function should read a register value and obtain the specific bits applying that mask. Here you don't test the bits as function header suggests. you just use the function to read the register value and apply the mask. We have for some older drivers the `read_masked` function implementation but we decided to not use it for newer drivers. so just read and apply mask here. (as in `adf4382_get_ref_div` )